examples/function_todomvc/src/main.rs

```diff
@@ -0,0 +1,257 @@
+use gloo::storage::{LocalStorage, Storage};
+use state::{Entry, Filter, State};
+use std::rc::Rc;
+use strum::IntoEnumIterator;
+use yew::{classes, function_component, html, use_effect_with_deps, use_reducer, Callback};
+
+mod components;
+mod state;
+
+use components::{
+    entry::Entry as EntryItem,
+    filter::Filter as FilterItem,
+    info_footer::InfoFooter,
+    input::Input,
+};
+
+pub enum Action {
+    Add(String),
+    Edit((usize, String)),
+    Remove(usize),
+    SetFilter(Filter),
+    ToggleAll,
+    ToggleEdit(usize),
+    Toggle(usize),
+    ClearCompleted,
+}
+
+const KEY: &str = "yew.functiontodomvc.self";
+
+#[function_component(App)]
+fn app() -> Html {
+    let state = use_reducer(
+        |prev: Rc<State>, action: Action| match action {
+            Action::Add(description) => {
+                let mut entries = prev.entries.clone();
+                entries.push(Entry {
+                    id: entries.last().map(|entry| entry.id + 1).unwrap_or(1),
+                    description,
+                    completed: false,
+                    editing: false,
+                });
+                State {
+                    entries,
+                    filter: prev.filter,
+                }
+            }
+            Action::Remove(id) => {
+                let mut entries = prev.entries.clone();
+                entries.retain(|entry| entry.id != id);
+                State {
+                    entries,
+                    filter: prev.filter,
+                }
+            }
+            Action::Toggle(id) => {
+                let mut entries = prev.entries.clone();
+                let entry = entries.iter_mut().find(|entry| entry.id == id);
+                if let Some(entry) = entry {
+                    entry.completed = !entry.completed;
+                }
+                State {
+                    entries,
+                    filter: prev.filter,
+                }
+            }
+            Action::Edit((id, description)) => {
+                let mut entries = prev.entries.clone();
+
+                if description.is_empty() {
+                    entries.retain(|entry| entry.id != id)
+                }
+
+                let entry = entries.iter_mut().find(|entry| entry.id == id);
+                if let Some(entry) = entry {
+                    entry.description = description;
+                    entry.editing = false;
+                }
+                State {
+                    entries,
+                    filter: prev.filter,
+                }
+            }
+            Action::ToggleEdit(id) => {
+                let mut entries = prev.entries.clone();
+                let entry = entries.iter_mut().find(|entry| entry.id == id);
+                if let Some(entry) = entry {
+                    entry.editing = !entry.editing;
+                }
+                State {
+                    entries,
+                    filter: prev.filter,
+                }
+            }
```

Comment:
I think this could be pushed down to the component, however, I think doing so would add a non trivial amount of complexities that might make it more difficult to follow, so not sure I want to go down this rabbit hole :)

As a fun aside I actually did go down this rabbit hole and came up with an interesting hook 🙃:

```rust
struct UseSwitchHandle {
    value: bool,
    switch: Rc<dyn Fn()>,
}

impl UseSwitchHandle {
    fn switch(self) {
        (self.switch)()
    }
}

impl Deref for UseSwitchHandle {
    type Target = bool;
    
    fn deref(&self) -> &Self::Target {
        &self.value
    }
}

fn use_switch(default: bool) -> UseSwitchHandle {
    use_hook(
        || default,
        move |hook, updater| {
            updater.post_render(move |state: &mut bool| {
                if *state != default {
                    *state = default;
                }
                false
            });
            let switch = Rc::new(move || {
                updater.callback(move |st: &mut bool| {
                    *st = !*st
                    true
                })
            });
            
            UseSwitchHandle {
                value: *hook,
                switch,
            }
        },
        |_| {},
    )
}
```

It will trigger a re-render once with the default switched over and after the render it switch the value back. I used this for the editing flag so by default it would be false and then on the double click I call the `switch` method and the component would re-render with editing == true but after render it would switch back silently so that on the next render (like when the props changed because the todo description has been updated) editing would be false :)
Thought it might interest you :)